src/descriptor/mod.rs

```diff
@@ -164,30 +189,34 @@ impl FromStr for DescriptorPublicKey {
                 .map_err(|_| {
                     DescriptorKeyParseError("Error while parsing master derivation path")
                 })?;
+            origin = Some((parent_fingerprint, origin_path));
+        }
 
-            let key_deriv = parts
+        let key_part = if origin == None {
+            Ok(s)
+        } else {
+            s.split(']')
+                .rev()
```

Comment:
Would this change accept incorrect strings with multiple `[` `]` as valid DescriptorPubkeys?

Reply:
No, a duplicated `[` would be counted as part of the size of the fingerprint: either the fingerprint is already valid (8 chars) and you get a "too long fingerprint, not 8 chars", or it eventually fails to parse [ as an hex char.

A duplicated `]`.... Would be accepted. Argh, nice catch correcting now. Tried with:
```rust
diff --git a/src/descriptor/mod.rs b/src/descriptor/mod.rs
index 91a0149..4eddec5 100644
--- a/src/descriptor/mod.rs
+++ b/src/descriptor/mod.rs
@@ -1474,6 +1474,14 @@ mod tests {
                 "Error while parsing simple public key"
             ))
         );
+
+        let desc = "[78412e3a]]03f28773c2d975288bc7d1d205c3748651b075fbc6610e58cddeeddf8f19405aa8";
+        assert_eq!(
+            DescriptorPublicKey::from_str(desc),
+            Err(DescriptorKeyParseError(
+                "Error while parsing simple public key"
+            ))
+        );
     }
 
     #[test]
```

Reply:
Fixed
```patch
diff --git a/src/descriptor/mod.rs b/src/descriptor/mod.rs
index 91a0149..af5d470 100644
--- a/src/descriptor/mod.rs
+++ b/src/descriptor/mod.rs
@@ -161,10 +161,11 @@ impl FromStr for DescriptorPublicKey {
             ));
         }
 
+        let mut parts = s[1..].split(']');
+
         // They may specify an origin
         let mut origin = None;
         if s.chars().next().unwrap() == '[' {
-            let mut parts = s[1..].split(']');
             let mut raw_origin = parts
                 .next()
                 .ok_or(DescriptorKeyParseError("Unclosed '['"))?
@@ -195,8 +196,7 @@ impl FromStr for DescriptorPublicKey {
         let key_part = if origin == None {
             Ok(s)
         } else {
-            s.split(']')
-                .rev()
+            parts
                 .next()
                 .ok_or(DescriptorKeyParseError("No key after origin."))
         }?;
@@ -1474,6 +1474,15 @@ mod tests {
                 "Error while parsing simple public key"
             ))
         );
+
+        // ..or invalid separators
+        let desc = "[78412e3a]]03f28773c2d975288bc7d1d205c3748651b075fbc6610e58cddeeddf8f19405aa8";
+        assert_eq!(
+            DescriptorPublicKey::from_str(desc),
+            Err(DescriptorKeyParseError(
+                "Error while parsing simple public key"
+            ))
+        );
     }
 
     #[test]
```